core/src/main/java/io/grpc/transport/MessageDeframer.java

```diff
@@ -139,8 +163,17 @@ public void request(int numMessages) {
     if (isClosed()) {
       return;
     }
-    pendingDeliveries += numMessages;
-    deliver();
+    // Make the number of requested messages visible to the transport thread.
+    transportCount.addAndGet(numMessages);
+    applicationCount += numMessages;
+    if (!appInRequest) {
+      appInRequest = true;
```

Comment:
Makes sense.  How do we "disallow" it? Just rely on documentation of the stream interface indicating that it is not thread safe?